### ISSUE TITLE
Bleadperl breaks ZEFRAM/Hash-SharedMem-0.003.tar.gz (with German locale)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14227]
Migrated from [rt.perl.org#123171](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123171) (status was 'resolved')

Searchable as RT123171$

**Comment by p5pRT** [comment]
### From @eserte

Hash\-SharedMem\-0\.003 started to fail with perl 5\.21\.x if a German locale is in effect \(and also probably for other non\-English locales\)\.

Pass/fail overview&#8203;: http&#8203;://matrix\.cpantesters\.org/?dist=Hash\-SharedMem\-0\.003
A sample fail report&#8203;: http&#8203;://www\.cpantesters\.org/cpan/report/09b60700\-684f\-11e4\-97e5\-c955e0bfc7aa

Regards\,
&nbsp;    Slaven

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Mon\, 10 Nov 2014 13&#8203;:03&#8203;:38 \-0800\, "slaven@&#8203;rezic\.de \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> said&#8203;:

&nbsp;  > Hash\-SharedMem\-0\.003 started to fail with perl 5\.21\.x if a German locale is in effect \(and also probably for other non\-English locales\)\.

&nbsp;  > Pass/fail overview&#8203;: http&#8203;://matrix\.cpantesters\.org/?dist=Hash\-SharedMem\-0\.003
&nbsp;  > A sample fail report&#8203;:
&nbsp;  > http&#8203;://www\.cpantesters\.org/cpan/report/09b60700\-684f\-11e4\-97e5\-c955e0bfc7aa

I cannot reproduce with recent bleadperls and German locale\. A PASS
report has been sent to cpantesters but has not yet arrived there\.

I'd say the bug here is foremost in the test itself which checks a
textual diagnostics message to match something like /No\\ such\\ file\\ or\\ directory/

There seem to be other factors involved than bleadperl\.

\-\- 
andreas

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @eserte

Dana Pon 10\. Stu 2014\, 18&#8203;:55&#8203;:19\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de reče&#8203;:
> >>>>> On Mon\, 10 Nov 2014 13&#8203;:03&#8203;:38 \-0800\, "slaven@&#8203;rezic\.de \(via RT\)"
> >>>>> \<perlbug\-followup@&#8203;perl\.org> said&#8203;:
> 
> > Hash\-SharedMem\-0\.003 started to fail with perl 5\.21\.x if a German
> > locale is in effect \(and also probably for other non\-English
> > locales\)\.
> 
> > Pass/fail overview&#8203;: http&#8203;://matrix\.cpantesters\.org/?dist=Hash\-
> > SharedMem\-0\.003
> > A sample fail report&#8203;:
> > http&#8203;://www\.cpantesters\.org/cpan/report/09b60700\-684f\-11e4\-97e5\-
> > c955e0bfc7aa
> 
> I cannot reproduce with recent bleadperls and German locale\. A PASS
> report has been sent to cpantesters but has not yet arrived there\.
> 
> I'd say the bug here is foremost in the test itself which checks a
> textual diagnostics message to match something like /No\\ such\\ file\\
> or\\ directory/
> 
> There seem to be other factors involved than bleadperl\.

The first failing test is using POSIX&#8203;::strerror\. This behaves different with bleadperl and maintperl\. On a FreeBSD 9 system&#8203;:

$ perl5\.21\.5 \-MPOSIX=strerror \-e 'warn strerror\(1\)'         
Operation not permitted at \-e line 1\.
$ perl5\.20\.1 \-MPOSIX=strerror \-e 'warn strerror\(1\)'
Operation nicht erlaubt at \-e line 1\.
$ locale
LANG=
LC\_CTYPE="de\_DE\.ISO8859\-1"
LC\_COLLATE="de\_DE\.ISO8859\-1"
LC\_TIME="de\_DE\.ISO8859\-1"
LC\_NUMERIC="de\_DE\.ISO8859\-1"
LC\_MONETARY="de\_DE\.ISO8859\-1"
LC\_MESSAGES="de\_DE\.ISO8859\-1"
LC\_ALL=de\_DE\.ISO8859\-1

On a Debian/wheezy system&#8203;:

$ env LC\_ALL=de\_DE\.utf\-8 perl5\.21\.5 \-MPOSIX=strerror \-e 'warn strerror\(1\)'
Operation not permitted at \-e line 1\.
$ env LC\_ALL=de\_DE\.utf\-8 perl5\.20\.1 \-MPOSIX=strerror \-e 'warn strerror\(1\)'
Die Operation ist nicht erlaubt at \-e line 1\.

**Comment by p5pRT** [comment]
### From @khwilliamson

On 11/10/2014 11&#8203;:28 PM\, slaven@&#8203;rezic\.de via RT wrote&#8203;:
> Dana Pon 10\. Stu 2014\, 18&#8203;:55&#8203;:19\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de reče&#8203;:
>>>>>>> On Mon\, 10 Nov 2014 13&#8203;:03&#8203;:38 \-0800\, "slaven@&#8203;rezic\.de \(via RT\)"
>>>>>>> \<perlbug\-followup@&#8203;perl\.org> said&#8203;:
>>
>>> Hash\-SharedMem\-0\.003 started to fail with perl 5\.21\.x if a German
>>> locale is in effect \(and also probably for other non\-English
>>> locales\)\.
>>
>>> Pass/fail overview&#8203;: http&#8203;://matrix\.cpantesters\.org/?dist=Hash\-
>>> SharedMem\-0\.003
>>> A sample fail report&#8203;:
>>> http&#8203;://www\.cpantesters\.org/cpan/report/09b60700\-684f\-11e4\-97e5\-
>>> c955e0bfc7aa
>>
>> I cannot reproduce with recent bleadperls and German locale\. A PASS
>> report has been sent to cpantesters but has not yet arrived there\.
>>
>> I'd say the bug here is foremost in the test itself which checks a
>> textual diagnostics message to match something like /No\\ such\\ file\\
>> or\\ directory/
>>
>> There seem to be other factors involved than bleadperl\.
>
> The first failing test is using POSIX&#8203;::strerror\. This behaves different with bleadperl and maintperl\. On a FreeBSD 9 system&#8203;:
>
> $ perl5\.21\.5 \-MPOSIX=strerror \-e 'warn strerror\(1\)'
> Operation not permitted at \-e line 1\.
> $ perl5\.20\.1 \-MPOSIX=strerror \-e 'warn strerror\(1\)'
> Operation nicht erlaubt at \-e line 1\.
> $ locale
> LANG=
> LC\_CTYPE="de\_DE\.ISO8859\-1"
> LC\_COLLATE="de\_DE\.ISO8859\-1"
> LC\_TIME="de\_DE\.ISO8859\-1"
> LC\_NUMERIC="de\_DE\.ISO8859\-1"
> LC\_MONETARY="de\_DE\.ISO8859\-1"
> LC\_MESSAGES="de\_DE\.ISO8859\-1"
> LC\_ALL=de\_DE\.ISO8859\-1
>
> On a Debian/wheezy system&#8203;:
>
> $ env LC\_ALL=de\_DE\.utf\-8 perl5\.21\.5 \-MPOSIX=strerror \-e 'warn strerror\(1\)'
> Operation not permitted at \-e line 1\.
> $ env LC\_ALL=de\_DE\.utf\-8 perl5\.20\.1 \-MPOSIX=strerror \-e 'warn strerror\(1\)'
> Die Operation ist nicht erlaubt at \-e line 1\.
>

This change was unintentional\, and I view it as a bug\.  The general rule 
is that Perl\-space code is not exposed to locale unless\, as long 
documented\, it is within the scope of 'use locale'\.  But calling C 
library functions \(through the POSIX module\) should act like the actual 
C library function\, with Perl interfering as little as possible\.  So 
strerror\(\) should do what it always has done\.  The change to blead is 
that "$\!" will not take locale into account outside the scope of "use 
locale"\, to go along with the general rule\.  However it turns out that 
POSIX&#8203;::strerror\(\) is implemented by calling "$\!"\, so this got 
inadvertently changed as a side\-effect\.

I propose fixing strerror to do what it did in maint\, unless there is a 
successful argument against it that persuades me or the pumpking\.

Perl can't do anything about the discrepancy in the German phrases on 
the two platforms\.

**Comment by p5pRT** [comment]
### From @khwilliamson

No comments\, so I've pushed the fix as
bbfbe7d903dc8287b167bd340a83daea4ef8c75b
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Dec 13 22&#8203;:05&#8203;:46 2014\, khw wrote&#8203;:
> No comments\, so I've pushed the fix as
> bbfbe7d903dc8287b167bd340a83daea4ef8c75b

I don’t see it yet\.  Did you actually push it?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Dec 13 22&#8203;:05&#8203;:46 2014\, khw wrote&#8203;:
> No comments\, so I've pushed the fix as
> bbfbe7d903dc8287b167bd340a83daea4ef8c75b

Actually cd9d242598e2\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @eserte

Dana Sub 13\. Pro 2014\, 22&#8203;:05&#8203;:46\, khw reče&#8203;:
> No comments\, so I've pushed the fix as
> bbfbe7d903dc8287b167bd340a83daea4ef8c75b

We have now a test failure if perl's test suite is run with LC\_ALL=de\_DE\.UTF\-8 or another non\-English locale\. With perl\-5\.21\.7&#8203;:

$ \./perl \-Ilib ext/POSIX/t/wrappers\.t
\.\.\.
not ok 39 \- strerror
\#   Failed test 'strerror'
\#   at ext/POSIX/t/wrappers\.t line 120\.
\#          got&#8203;: 'Datei oder Verzeichnis nicht gefunden'
\#     expected&#8203;: 'No such file or directory'
\.\.\.

**Comment by p5pRT** [comment]
### From @eserte

Dana Pon 29\. Pro 2014\, 09&#8203;:45&#8203;:39\, slaven@&#8203;rezic\.de reče&#8203;:
> Dana Sub 13\. Pro 2014\, 22&#8203;:05&#8203;:46\, khw reče&#8203;:
> > No comments\, so I've pushed the fix as
> > bbfbe7d903dc8287b167bd340a83daea4ef8c75b
> 
> We have now a test failure if perl's test suite is run with
> LC\_ALL=de\_DE\.UTF\-8 or another non\-English locale\. With perl\-5\.21\.7&#8203;:
> 
> $ \./perl \-Ilib ext/POSIX/t/wrappers\.t
> \.\.\.
> not ok 39 \- strerror
> \#   Failed test 'strerror'
> \#   at ext/POSIX/t/wrappers\.t line 120\.
> \#          got&#8203;: 'Datei oder Verzeichnis nicht gefunden'
> \#     expected&#8203;: 'No such file or directory'
> \.\.\.

Possible fix attached\.

**Comment by p5pRT** [comment]
### From @eserte

<details><summary>0001-test-POSIX-strerror-with-C-and-current-possibly-non-.patch</summary>

```diff
From d6e594d43a815ce6bf3d879ae5377a4868c8cb2b Mon Sep 17 00:00:00 2001
From: Slaven Rezic <slaven@rezic.de>
Date: Mon, 29 Dec 2014 18:53:29 +0100
Subject: [PATCH] test POSIX::strerror() with C and current (possibly non-C)
 locale

For: RT #123171
---
 ext/POSIX/t/wrappers.t | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/ext/POSIX/t/wrappers.t b/ext/POSIX/t/wrappers.t
index 5b9f6d6..dc02601 100644
--- a/ext/POSIX/t/wrappers.t
+++ b/ext/POSIX/t/wrappers.t
@@ -114,10 +114,25 @@ is(POSIX::sleep(0), 0, 'sleep');
 is(POSIX::sprintf('%o', 42), '52', 'sprintf');
 is(POSIX::sqrt(256), 16, 'sqrt');
 is_deeply([POSIX::stat($temp_file)], [stat $temp_file], 'stat');
+
 {
     local $! = 2;
+    my $error;
+    {
+	use locale;
+	$error = "$!";
+    }
+    is(POSIX::strerror(2), $error, 'strerror under current locale');
+}
+
+SKIP: {
+    skip("no setlocale available", 1) unless $Config{d_setlocale};
+    local $! = 2;
     my $error = "$!";
-    is(POSIX::strerror(2), $error, 'strerror');
+    my $lc = POSIX::setlocale(POSIX::LC_ALL());
+    POSIX::setlocale(POSIX::LC_ALL(), 'C');
+    is(POSIX::strerror(2), $error, 'strerror under C locale');
+    POSIX::setlocale(POSIX::LC_ALL(), $lc);
 }
 
 is(POSIX::strstr('BBFRPRAFPGHPP', 'FP'), 7, 'strstr');
-- 
2.1.2


```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

On Mon Dec 29 09&#8203;:55&#8203;:42 2014\, slaven@&#8203;rezic\.de wrote&#8203;:
> Dana Pon 29\. Pro 2014\, 09&#8203;:45&#8203;:39\, slaven@&#8203;rezic\.de reče&#8203;:
> > Dana Sub 13\. Pro 2014\, 22&#8203;:05&#8203;:46\, khw reče&#8203;:
> > > No comments\, so I've pushed the fix as
> > > bbfbe7d903dc8287b167bd340a83daea4ef8c75b
> > 
> > We have now a test failure if perl's test suite is run with
> > LC\_ALL=de\_DE\.UTF\-8 or another non\-English locale\. With perl\-5\.21\.7&#8203;:
> > 
> > $ \./perl \-Ilib ext/POSIX/t/wrappers\.t
> > \.\.\.
> > not ok 39 \- strerror
> > \#   Failed test 'strerror'
> > \#   at ext/POSIX/t/wrappers\.t line 120\.
> > \#          got&#8203;: 'Datei oder Verzeichnis nicht gefunden'
> > \#     expected&#8203;: 'No such file or directory'
> > \.\.\.
> 
> Possible fix attached\.

I used part of your fix in commit adf06ddfd8b6d9835479092f95ab172a8d5a3a87\,
but I came up with a way to more likely test on a non\-English locale than your patch did\.  Anyway\, the problem was just the \.t\, and not the actual code it was testing
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Mon Dec 29 09&#8203;:55&#8203;:42 2014\, slaven@&#8203;rezic\.de wrote&#8203;:
> Dana Pon 29\. Pro 2014\, 09&#8203;:45&#8203;:39\, slaven@&#8203;rezic\.de reče&#8203;:
> > Dana Sub 13\. Pro 2014\, 22&#8203;:05&#8203;:46\, khw reče&#8203;:
> > > No comments\, so I've pushed the fix as
> > > bbfbe7d903dc8287b167bd340a83daea4ef8c75b
> > 
> > We have now a test failure if perl's test suite is run with
> > LC\_ALL=de\_DE\.UTF\-8 or another non\-English locale\. With perl\-5\.21\.7&#8203;:
> > 
> > $ \./perl \-Ilib ext/POSIX/t/wrappers\.t
> > \.\.\.
> > not ok 39 \- strerror
> > \#   Failed test 'strerror'
> > \#   at ext/POSIX/t/wrappers\.t line 120\.
> > \#          got&#8203;: 'Datei oder Verzeichnis nicht gefunden'
> > \#     expected&#8203;: 'No such file or directory'
> > \.\.\.
> 
> Possible fix attached\.

I used part of your fix in commit adf06ddfd8b6d9835479092f95ab172a8d5a3a87\,
but I came up with a way to more likely test on a non\-English locale than your patch did\.  Anyway\, the problem was just the \.t\, and not the actual code it was testing
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
### From @steve-m-hay

On 7 January 2015 at 04&#8203;:44\, Karl Williamson via RT
\<perlbug\-comment@&#8203;perl\.org> wrote&#8203;:
> On Mon Dec 29 09&#8203;:55&#8203;:42 2014\, slaven@&#8203;rezic\.de wrote&#8203;:
>> Dana Pon 29\. Pro 2014\, 09&#8203;:45&#8203;:39\, slaven@&#8203;rezic\.de reče&#8203;:
>> > Dana Sub 13\. Pro 2014\, 22&#8203;:05&#8203;:46\, khw reče&#8203;:
>> > > No comments\, so I've pushed the fix as
>> > > bbfbe7d903dc8287b167bd340a83daea4ef8c75b
>> >
>> > We have now a test failure if perl's test suite is run with
>> > LC\_ALL=de\_DE\.UTF\-8 or another non\-English locale\. With perl\-5\.21\.7&#8203;:
>> >
>> > $ \./perl \-Ilib ext/POSIX/t/wrappers\.t
>> > \.\.\.
>> > not ok 39 \- strerror
>> > \#   Failed test 'strerror'
>> > \#   at ext/POSIX/t/wrappers\.t line 120\.
>> > \#          got&#8203;: 'Datei oder Verzeichnis nicht gefunden'
>> > \#     expected&#8203;: 'No such file or directory'
>> > \.\.\.
>>
>> Possible fix attached\.
>
> I used part of your fix in commit adf06ddfd8b6d9835479092f95ab172a8d5a3a87\,
> but I came up with a way to more likely test on a non\-English locale than your patch did\.  Anyway\, the problem was just the \.t\, and not the actual code it was testing

I have a new test failure on Windows since commit
adf06ddfd8b6d9835479092f95ab172a8d5a3a87&#8203;:

C&#8203;:\\Dev\\Git\\perl\\t>\.\\perl harness \-v \.\.\\ext\\POSIX\\t\\wrappers\.t
\.\./ext/POSIX/t/wrappers\.t \.\. Your vendor has not defined POSIX macro LC\_MESSAGES
\, used at t/wrappers\.t line 30
Dubious\, test returned 1 \(wstat 256\, 0x100\)
No subtests run

Test Summary Report


---
\.\./ext/POSIX/t/wrappers\.t \(Wstat&#8203;: 256 Tests&#8203;: 0 Failed&#8203;: 0\)
&nbsp;  Non\-zero exit status&#8203;: 1
&nbsp;  Parse errors&#8203;: No plan found in TAP output
Files=1\, Tests=0\,  1 wallclock secs \( 0\.02 usr \+  0\.00 sys =  0\.02 CPU\)
Result&#8203;: FAIL

**Comment by p5pRT** [comment]
### From @khwilliamson

On 01/07/2015 01&#8203;:31 AM\, Steve Hay wrote&#8203;:
> On 7 January 2015 at 04&#8203;:44\, Karl Williamson via RT
> \<perlbug\-comment@&#8203;perl\.org> wrote&#8203;:
>> On Mon Dec 29 09&#8203;:55&#8203;:42 2014\, slaven@&#8203;rezic\.de wrote&#8203;:
>>> Dana Pon 29\. Pro 2014\, 09&#8203;:45&#8203;:39\, slaven@&#8203;rezic\.de reče&#8203;:
>>>> Dana Sub 13\. Pro 2014\, 22&#8203;:05&#8203;:46\, khw reče&#8203;:
>>>>> No comments\, so I've pushed the fix as
>>>>> bbfbe7d903dc8287b167bd340a83daea4ef8c75b
>>>>
>>>> We have now a test failure if perl's test suite is run with
>>>> LC\_ALL=de\_DE\.UTF\-8 or another non\-English locale\. With perl\-5\.21\.7&#8203;:
>>>>
>>>> $ \./perl \-Ilib ext/POSIX/t/wrappers\.t
>>>> \.\.\.
>>>> not ok 39 \- strerror
>>>> \#   Failed test 'strerror'
>>>> \#   at ext/POSIX/t/wrappers\.t line 120\.
>>>> \#          got&#8203;: 'Datei oder Verzeichnis nicht gefunden'
>>>> \#     expected&#8203;: 'No such file or directory'
>>>> \.\.\.
>>>
>>> Possible fix attached\.
>>
>> I used part of your fix in commit adf06ddfd8b6d9835479092f95ab172a8d5a3a87\,
>> but I came up with a way to more likely test on a non\-English locale than your patch did\.  Anyway\, the problem was just the \.t\, and not the actual code it was testing
>
> I have a new test failure on Windows since commit
> adf06ddfd8b6d9835479092f95ab172a8d5a3a87&#8203;:
>
> C&#8203;:\\Dev\\Git\\perl\\t>\.\\perl harness \-v \.\.\\ext\\POSIX\\t\\wrappers\.t
> \.\./ext/POSIX/t/wrappers\.t \.\. Your vendor has not defined POSIX macro LC\_MESSAGES
> \, used at t/wrappers\.t line 30
> Dubious\, test returned 1 \(wstat 256\, 0x100\)
> No subtests run

This should be fixed by ed0cac5b003824d9b42c00752f4f894700e967d4

But I have a couple questions\.  I was trying this out on dromedary\, and 
it appears that strerror\(\) on Windows always returns in English\.  That 
seems odd\.

I also found that a SIG\_WARN handler wasn't being properly called there\. 
&nbsp;  Is that to be expected?
>
> Test Summary Report
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \.\./ext/POSIX/t/wrappers\.t \(Wstat&#8203;: 256 Tests&#8203;: 0 Failed&#8203;: 0\)
>    Non\-zero exit status&#8203;: 1
>    Parse errors&#8203;: No plan found in TAP output
> Files=1\, Tests=0\,  1 wallclock secs \( 0\.02 usr \+  0\.00 sys =  0\.02 CPU\)
> Result&#8203;: FAIL
>

**Comment by p5pRT** [comment]
### From @jandubois

On Wed\, Jan 7\, 2015 at 10&#8203;:50 AM\, Karl Williamson
\<public@&#8203;khwilliamson\.com> wrote&#8203;:
> It appears that strerror\(\) on Windows always returns in English\.

I believe that is true; there are no localized versions of MSVCRT
\(AFAIK\)\.  The system message catalogs are localized though\, so $\!
should always be English\, but $^E should be in local \*system\* language
\(but still not overridable via LANG\, or LC\_\*\)\.

Cheers\,
\-Jan

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
